cpp/src/strings/regex/regcomp.cpp

```diff
@@ -459,16 +459,40 @@ class regex_parser {
     }
 
     // The quantifiers require at least one "real" previous item.
-    // We are throwing an error in these two if-checks for invalid quantifiers.
+    // We are throwing an errors for invalid quantifiers.
     // Another option is to just return CHAR silently here which effectively
     // treats the chr character as a literal instead as a quantifier.
     // This could lead to confusion where sometimes unescaped quantifier characters
     // are treated as regex expressions and sometimes they are not.
     if (_items.empty()) { CUDF_FAIL("invalid regex pattern: nothing to repeat at position 0"); }
 
+    // Check that the previous item can be used with quantifiers.
+    // If the previous item is a capture group, we need to check items inside the
+    // capture group can be used with quantifiers too.
+    // (Note that capture groups can be nested).
+    auto previous_type = _items.back().type;
+    if (previous_type == RBRA) {  // previous item is a capture group
+      // look for matching LBRA
+      auto nested_count = 0;
+      auto lbra_itr =
+        std::find_if(_items.rbegin(), _items.rend(), [nested_count](auto const& item) mutable {
+          nested_count += (item.type == RBRA);
+          return (item.type == LBRA || item.type == LBRA_NC) && (nested_count-- == 0);
```

Comment:
The control flow looks a bit hidden to me - short-circuiting plus post-decrement in a condition. How about separating the nesting scan from the condition?
```suggestion
          auto is_closing = item.type == RBRA;
          auto is_opening = item.type == LBRA || item.type == LBRA_NC;
          nested_count += is_closing - is_opening;
          return is_opening && nested_count == -1; // look for the first opening parenthesis that starts outside this context
```
You could also initialize nested_count to 1, since you start after a parenthesis, then the condition would look less surprising?